xsdvalidate/xsd_validate_activity.go

```diff
@@ -0,0 +1,192 @@
+
+	// Pattern is the regular expression for files to validate.
+	Pattern string
+
+	// XSDFilePathOrURL is the path of the XSD file to use for validation.
+	XSDFilePathOrURL string
+}
+
+type XSDValidateActivityResult struct {
+	Failed []string
+}
+
+// Execute checks all files, in a given directory, matching a certain pattern.
+func (a *XSDValidateActivity) Execute(
+	ctx context.Context,
+	params *XSDValidateActivityParams,
+) (*XSDValidateActivityResult, error) {
+	logger := temporal.GetLogger(ctx)
+	logger.V(1).Info("Executing XMLValidateActivity", "DirectoryPath", params.DirectoryPath)
+
+	// Check is XSD has been provided as a file, otherwise attempt download.
+	_, err := os.Stat(params.XSDFilePathOrURL)
+	if errors.Is(err, os.ErrNotExist) {
+		// Attempt to download XSD file.
+		file, err := os.CreateTemp("", "xsdValidatrActvitity")
+		if err != nil {
+			return nil, err
+		}
+		defer os.Remove(file.Name())
+
+		err = DownloadXsdFile(params.XSDFilePathOrURL, file.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		params.XSDFilePathOrURL = file.Name()
+	}
+
+	// Find matching files.
+	var filesFound []string
+
+	filesFound, err = MatchingFilesRelativeToDirectory(params.DirectoryPath, params.Pattern)
+	if err != nil {
+		return nil, fmt.Errorf("XSDValidateActivity match files: %v", err)
+	}
+
+	// Prepare XSD handler.
+	err = xsdvalidator.Init()
+	if err != nil {
+		return nil, fmt.Errorf("XSDValidateActivity xsd init error: %s", err)
+	}
+	defer xsdvalidator.Cleanup()
+
+	// Note: NewXsdHandlerUrl doesn't handle https which is why DownloadXsdFile is used
+	// (https://github.com/terminalstatic/go-xsd-validate/issues/11)
+	xsdHandler, err := xsdvalidator.NewXsdHandlerUrl(
+		params.XSDFilePathOrURL,
+		xsdvalidator.ParsErrDefault,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("XSDValidateActivity new xsd handler for %v: %s", err, params.XSDFilePathOrURL)
+	}
+
+	// Check matched files.
+	var failed []string
+
+	for _, matchedFile := range filesFound {
+		checkFilePath := filepath.Join(params.DirectoryPath, matchedFile)
+
+		// Read XML file contents.
+		xmlFile, err := os.Open(filepath.Clean(checkFilePath))
+		if err != nil {
+			return nil, err
+		}
+		defer xmlFile.Close()
+
+		xmlData, err := io.ReadAll(xmlFile)
+		if err != nil {
+			return nil, err
+		}
+
+		// Prepare XML handler for file contents.
+		xmlhandler, err := xsdvalidator.NewXmlHandlerMem(xmlData, xsdvalidator.ParsErrDefault)
+		if err != nil {
+			return nil, err
+		}
+		defer xmlhandler.Free()
+
+		// Validate the XML against the XSD.
+		err = xsdHandler.Validate(xmlhandler, xsdvalidator.ValidErrDefault)
+		if err != nil {
+			failed = append(failed, matchedFile)
+		}
+	}
+
+	return &XSDValidateActivityResult{Failed: failed}, nil
+}
+
+func DownloadXsdFile(url string, xsdFilePath string) error {
+	// Make GET request.
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := http.DefaultClient.Do(req)
```

Comment:
[Avoid the http.DefaultClient](https://artefactual.dev/docs/programming-languages/go/http/#avoid-the-default-client). :)

Reply:
Fixed!